### PR TITLE
[api] Run mass upgrades from API #41

### LINKED ISSUE #41: [api] Run mass upgrades from API
Ideally having an endpoint like `/api/build/<pk>/massupgrade/`.

### PR DESCRIPTION
**Pull request by PabloCastellano**
Closes #41

### CODE REVIEW

File: openwisp_firmware_upgrader/admin.py [hunk 1]
```diff
@@ -76,7 +76,7 @@ def upgrade_selected(self, request, queryset):
             # returning None will display the change list page again
             return None
         upgrade_all = request.POST.get('upgrade_all')
-        upgrade_related = request.POST.get('upgrade_related')
+        upgrade_related = request.POST.get('upgrade_related')  # UNUSED!!
```
Comment (nemesisdesign) [inline openwisp_firmware_upgrader/admin.py:79]: is this not being used at all there?

Comment (PabloCastellano) [reply openwisp_firmware_upgrader/admin.py:80]: I don't see how it's used:

This is the logic simplified:

```
upgrade_all = request.POST.get('upgrade_all')
upgrade_related = request.POST.get('upgrade_related') 
if upgrade_all or upgrade_related:
    batch = build.batch_upgrade(firmwareless=upgrade_all)
```

Should I remove it?

Comment (PabloCastellano) [reply openwisp_firmware_upgrader/admin.py:86]: I've opened an issue to track it -> https://github.com/openwisp/openwisp-firmware-upgrader/issues/81

File: openwisp_firmware_upgrader/api/views.py [hunk 2]
```diff
@@ -67,6 +68,20 @@ class BuildDetailView(OrgAPIMixin, generics.RetrieveUpdateDestroyAPIView):
...
 
+class BuildMassUpgradeView(OrgAPIMixin, generics.GenericAPIView):
+    model = Build
+    queryset = Build.objects.all().select_related('category')
+    serializer_class = BuildSerializer
+    lookup_fields = ['pk']
+    organization_field = 'category__organization'
+
+    def post(self, request, pk):
+        upgrade_all = request.POST.get('upgrade_all')
```
Comment (nemesisdesign) [inline openwisp_firmware_upgrader/api/views.py:79]: I suggest to make this line more explicit:
```python
upgrade_all = request.POST.get('upgrade_all') is not None
```
So we get either `True` or `False`

File: openwisp_firmware_upgrader/tests/test_api.py [hunk 3]
```diff
@@ -220,6 +220,27 @@ def test_build_delete(self):
         self.assertEqual(r.status_code, 204)
         self.assertEqual(Build.objects.count(), 0)
 
+    def test_build_mass_upgrade(self):
+        build = self._create_build()
+        self.assertEqual(BatchUpgradeOperation.objects.count(), 0)
+        self.assertEqual(DeviceFirmware.objects.count(), 0)
+        with self.subTest("Existing build"):
```
Comment (nemesisdesign) [inline openwisp_firmware_upgrader/tests/test_api.py:227]: try to be consistent with the quoting style used

File: openwisp_firmware_upgrader/tests/test_api.py [hunk 4]
```diff
@@ -220,6 +220,27 @@ def test_build_delete(self):
...
+            url = reverse('upgrader:api_build_mass_upgrade', args=[build.pk])
+            with self.assertNumQueries(7):
+                r = self.client.post(url)
+            self.assertEqual(BatchUpgradeOperation.objects.count(), 1)
+            self.assertEqual(DeviceFirmware.objects.count(), 0)
+            batch = BatchUpgradeOperation.objects.first()
+            self.assertEqual(r.status_code, 201)
+            self.assertEqual(r.data, {"batch": str(batch.pk)})
```
Comment (nemesisdesign) [inline openwisp_firmware_upgrader/tests/test_api.py:235]: here as well

File: openwisp_firmware_upgrader/tests/test_api.py [hunk 5]
```diff
@@ -220,6 +220,27 @@ def test_build_delete(self):
...
+
+        with self.subTest("Non existing build"):
```
Comment (nemesisdesign) [inline openwisp_firmware_upgrader/tests/test_api.py:237]: here as well

File: openwisp_firmware_upgrader/tests/test_api.py [hunk 6]
```diff
@@ -220,6 +220,27 @@ def test_build_delete(self):
...
+            url = reverse('upgrader:api_build_mass_upgrade', args=[uuid.uuid4()])
+            with self.assertNumQueries(4):
+                r = self.client.post(url)
+            self.assertEqual(r.status_code, 404)
+            self.assertEqual(r.json(), {"detail": "Not found."})
```
Comment (nemesisdesign) [inline openwisp_firmware_upgrader/tests/test_api.py:242]: here as well

File: openwisp_firmware_upgrader/tests/test_api.py (hunk 5 above)
Comment (nemesisdesign) [reply openwisp_firmware_upgrader/tests/test_api.py:237]: PS: I would make a dedicated test for this as in the other endpoints

**Pull request merged**